### PR TITLE
tty,win: Fix duplicate vt100 fn key

### PR DESCRIPTION
**Pull request by erw7**
By #2114 function keys etc duplicate in neovim. I think that this is the correct condition.

### CODE REVIEW

**Comment by erw7** [comment]
@bzoz Which of the following actions you saw?

```
2 '\u001ba'
2 '\u001ba'
```

```
3 'u001baa'
```

If it is the former, will you delay the key repeat and show it? If it is the latter, it means that the following event has occurred with "a" key up or "alt" key up.

```
KEV.wVirtualKeyCode == VK_MENU
KEV.uChar.UnicodeChar == 0x41(a)
```

I can not believe it.

**Comment by bzoz** [comment]
A test, great!

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1190/

**Comment by bzoz** [comment]
Sorry, I'll be away for a week and cannot test this. I'll rubber stamp LGTM, the change makes sense, looks good and the test is great.

Can someone confirm that this works?

**Comment by joaocgreis** [comment]
This issue is very noticeable moving around in the repl. I tried this change and it appears to fix the issue.

File: test/test-tty-duplicate-key.c [hunk 1]
```diff
@@ -0,0 +1,313 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (richardlau) [inline test/test-tty-duplicate-key.c]: For new files, this should be:
```C
/* Copyright libuv project contributors. All rights reserved.
```

unless this test file is heavily based on an existing test file.

File: test/test-tty-duplicate-key.c [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    kb_state[VK_SHIFT] = 0x01;
+  }
+  ret = ToUnicode(virt_key, KEV(0).wVirtualScanCode, kb_state, buf, 2, 0);
+  if (ret == 1) {
+    if(!is_wsl &&
+        ((ctr_key_state & LEFT_ALT_PRESSED) ||
+         (ctr_key_state & RIGHT_ALT_PRESSED))) {
+      /*
+       * If ALT key is pressed, the UnicodeChar value of the keyup event is
+       * set to 0 on nomal console. Emulate this behaviro.
```
Comment (bzoz) [inline test/test-tty-duplicate-key.c]: nit: spelling, behavior

File: test/test-tty-duplicate-key.c [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+       * See https://github.com/Microsoft/console/issues/320
+       */
+      KEV(0).uChar.UnicodeChar = buf[0];
+      KEV(1).uChar.UnicodeChar = 0;
+    } else{
+      /*
+       * In WSL UnicodeChar is normally set. This behaviro cause #2111.
```
Comment (bzoz) [inline test/test-tty-duplicate-key.c]: same here

**Comment by bzoz** [comment]
Without the change, two of the three new tests pass:
```
ok 301 - tty_composing_character
ok 302 - tty_duplicate_alt_modifier_key
not ok 303 - tty_duplicate_vt100_fn_key
# exit code 3
# Output from process `tty_duplicate_vt100_fn_key`:
# expected nread 4, but found 8
# expect 0x1b 0x5b 0x5b 0x41 , but found 0x1b 0x5b 0x5b 0x41 0x1b 0x5b 0x5b 0x41
# Assertion failed in test-tty-duplicate-key.c on line 66: FALSE
```
When I've made that change I've only checked Alt+something, not other special keys. This is also why originally I did not see any change with and without this change.

Anyway, @cjihrig, this probably needs a new libuv release, since currently Node is broken in a rather spectacular way.

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1206/

File: test/test-tty-duplicate-key.c [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+static void dump_str(const char* str, ssize_t len) {
+  ssize_t i;
+  for (i = 0; i < len; i++) {
+    fprintf(stderr, "%#02x ", *(str + i));
+  }
+}
+
+static void print_err_msg(const char* expect, ssize_t expect_len,
+    const char* found, ssize_t found_len) {
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Can you line these up vertically with the arguments on the previous line.

File: test/test-tty-duplicate-key.c [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  fprintf(stderr, "expect ");
+  dump_str(expect, expect_len);
+  fprintf(stderr, ", but found ");
+  dump_str(found, found_len);
+  fprintf(stderr, "\n");
+}
+
+static void tty_alloc(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
+  buf->base = malloc(size);
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Maybe assert that `buf->base` is not `NULL`.

File: test/test-tty-duplicate-key.c [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  buf->len = size;
+}
+
+static void tty_read(uv_stream_t* tty_in, ssize_t nread, const uv_buf_t* buf) {
+  if (nread > 0) {
+    if (nread != expect_nread) {
+      fprintf(stderr, "expected nread %ld, but found %ld\n",
+          (long)expect_nread, (long)nread);
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Please line this up vertically with the previous line's arguments.

File: test/test-tty-duplicate-key.c [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+      ASSERT(FALSE);
+    }
+    uv_close((uv_handle_t*) tty_in, NULL);
+  } else {
+    ASSERT(nread == 0);
+  }
+}
+
+static void make_key_event_records(WORD virt_key, DWORD ctr_key_state,
+    BOOL is_wsl, INPUT_RECORD* records) {
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Same comment about lining up the function arguments. I won't keep pointing it out.

File: test/test-tty-duplicate-key.c [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    KEV(0).uChar.UnicodeChar = KEV(1).uChar.UnicodeChar = 0;
+  }
+# undef KEV
+}
+
+TEST_IMPL(tty_duplicate_vt100_fn_key) {
+  int r;
+  int ttyin_fd;
+  uv_tty_t tty_in;
+  uv_loop_t* loop = uv_default_loop();
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Can you separate the assignment from the declaration.

File: test/test-tty-duplicate-key.c [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+  ASSERT(written == ARRAY_SIZE(records));
+  make_key_event_records(VK_NUMPAD2, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  make_key_event_records(VK_NUMPAD8, LEFT_ALT_PRESSED, FALSE, records);
+  WriteConsoleInputW(handle, records, ARRAY_SIZE(records), &written);
+  ASSERT(written == ARRAY_SIZE(records));
+  WriteConsoleInputW(handle, &alt_records[1], 1, &written);
+
+  Sleep(100);
```
Comment (cjihrig) [inline test/test-tty-duplicate-key.c]: Is it possible to write these tests without `Sleep()`?

Comment (erw7) [reply test/test-tty-duplicate-key.c]: I removed the `Sleep()` call for the time being. Although there is no problem in the environment at hand, I think there is a possibility that problems will arise in a fast environment.

File: test/test-tty-duplicate-key.c (hunk 5 above)
Comment (erw7) [reply test/test-tty-duplicate-key.c]: I fixed it, but as far as I see https://github.com/libuv/libuv/pull/2177#issuecomment-460554963, https://github.com/libuv/libuv/pull/2177#issuecomment-461870631 it was better not to fix it at the moment?

Comment (cjihrig) [reply test/test-tty-duplicate-key.c]: I think it's different when the code is being written for the first time versus churning over existing code.

**Comment by cjihrig** [comment]
CI with no related failures: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1216/

**Comment by cjihrig** [comment]
Landed in https://github.com/libuv/libuv/commit/7ed1eced3145c9eb17f831c2d66cdaa3d440d904. Thanks!

**Pull request closed without merging**